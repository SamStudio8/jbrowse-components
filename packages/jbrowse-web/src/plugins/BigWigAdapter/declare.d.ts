declare module 'abortable-promise-cache'
declare module '@gmod/jbrowse-core/util/io'
