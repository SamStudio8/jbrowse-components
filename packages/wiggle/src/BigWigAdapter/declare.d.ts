declare module 'abortable-promise-cache'
declare module '@gmod/jbrowse-core/util/QuickLRU'
