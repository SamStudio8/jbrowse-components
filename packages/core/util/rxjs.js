import { Observable } from 'rxjs'

/**
 * Wrapper for rxjs Observable.create with improved error handling
 * @param {function} func observer function, could be async
 */
export function ObservableCreate(func) {
  return Observable.create(observer => {
    try {
      const ret = func(observer)
      // catch async errors
      if (ret && ret.catch) ret.catch(error => observer.error(error))
      return ret
    } catch (error) {
      // catch sync errors
      observer.error(error)
    }
    return undefined
  })
}
