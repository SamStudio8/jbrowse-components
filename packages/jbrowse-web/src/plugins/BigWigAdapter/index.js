import Plugin from '@gmod/jbrowse-core/Plugin'
import AdapterType from '@gmod/jbrowse-core/pluggableElementTypes/AdapterType'
import { ConfigurationSchema } from '@gmod/jbrowse-core/configuration'
import AdapterClass from './BigWigAdapter'

const configSchema = ConfigurationSchema(
  'BigWigAdapter',
  {
    bigWigLocation: {
      type: 'fileLocation',
      defaultValue: { uri: '/path/to/my.bw' },
    },
  },
  { explicitlyTyped: true },
)

export default class BigWigAdapterPlugin extends Plugin {
  install(pluginManager) {
    pluginManager.addAdapterType(
      () =>
        new AdapterType({
          name: 'BigWigAdapter',
          configSchema,
          AdapterClass,
        }),
    )
  }
}
