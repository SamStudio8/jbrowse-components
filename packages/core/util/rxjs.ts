import { Observable, Observer } from 'rxjs'

/**
 * Wrapper for rxjs Observable.create with improved error handling
 * @param {function} func observer function, could be async
 */
export function ObservableCreate<T>(func:Function):Observable<T> {
  return Observable.create((observer:Observer<T>) => {
    try {
      const ret = func(observer)
      // catch async errors
      if (ret && ret.catch) ret.catch((error:any) => observer.error(error))
      return ret
    } catch (error) {
      // catch sync errors
      observer.error(error)
    }
    return undefined
  })
}

export function removeme() {}
