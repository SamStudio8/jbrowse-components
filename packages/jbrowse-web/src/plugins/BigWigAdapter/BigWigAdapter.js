import { BigWig } from '@gmod/bbi'

import BaseAdapter from '@gmod/jbrowse-core/BaseAdapter'
import { openLocation } from '@gmod/jbrowse-core/util/io'
import SimpleFeature from '@gmod/jbrowse-core/util/simpleFeature'
import { ObservableCreate } from '@gmod/jbrowse-core/util/rxjs'

export default class BigWigAdapter extends BaseAdapter {
  static capabilities = ['getFeatures', 'getRefNames']

  constructor(config) {
    super(config)
    this.bigwig = new BigWig({
      filehandle: openLocation(config.bigWigLocation),
    })
  }

  async getRefNames() {
    const header = await this.bigwig.getHeader()
    return Object.keys(header.refsByName)
  }

  async refIdToName(refId) {
    return ((await this.bigwig.getHeader()).refsByNumber[refId] || {}).name
  }

  async getGlobalStats() {
    const header = await this.bigwig.getHeader()
    return header.totalSummary
  }

  /**
   * Fetch features for a certain region
   * @param {Region} param
   * @returns {Observable[Feature]} Observable of Feature objects in the region
   */
  getFeatures({ /* assembly, */ refName, start, end }) {
    return ObservableCreate(async observer => {
      const records = await this.bigwig.getFeatures(refName, start, end)
      records.forEach(record => {
        observer.next(
          new SimpleFeature({
            id: record.start + 1,
            data: record,
          }),
        )
      })
      observer.complete()
    })
  }
}
