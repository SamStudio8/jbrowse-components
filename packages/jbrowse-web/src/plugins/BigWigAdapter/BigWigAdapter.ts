import { BigWig } from '@gmod/bbi'
import AbortablePromiseCache from 'abortable-promise-cache'
import QuickLRU from 'quick-lru'
import { Observable,from } from 'rxjs'
import { mergeMap,concatAll, map } from 'rxjs/operators'

import BaseAdapter, {Region} from '@gmod/jbrowse-core/BaseAdapter'
import { openLocation } from '@gmod/jbrowse-core/util/io'
import SimpleFeature from '@gmod/jbrowse-core/util/simpleFeature'
import { ObservableCreate } from '@gmod/jbrowse-core/util/rxjs'

import { rectifyStats, scoresToStats } from './util'

export default class BigWigAdapter extends BaseAdapter {
  private bigwig: BigWig
  private statsCache: AbortablePromiseCache

  static capabilities = ['getFeatures', 'getRefNames']

  constructor(config) {
    super(config)
    this.bigwig = new BigWig({
      filehandle: openLocation(config.bigWigLocation),
    })
    const bigwigRef = this.bigwig
    this.statsCache = new AbortablePromiseCache({
      cache: new QuickLRU({ maxSize: 1000 }),
      async fill(region, abortSignal) {
        const { refName, start, end, bpPerPx } = region
        const feats = await bigwigRef.getFeatures(refName, start, end, {
          signal: abortSignal,
          basesPerSpan: (bpPerPx * 20) / Math.log(end - start),
        })
        return scoresToStats(region, feats)
      },
    })
  }

  async getRefNames() {
    const header = await this.bigwig.getHeader()
    return Object.keys(header.refsByName)
  }

  async refIdToName(refId) {
    return ((await this.bigwig.getHeader()).refsByNumber[refId] || {}).name
  }

  async getGlobalStats(abortSignal) {
    const header = await this.bigwig.getHeader(abortSignal)
    return header.totalSummary
  }

  // todo: incorporate summary blocks
  getRegionStats(region, abortSignal) {
    const { refName, start, end, bpPerPx } = region
    return this.statsCache.get(
      `${refName}_${start}_${end}_${bpPerPx}`,
      region,
      abortSignal,
    )
  }

  // todo: add caching
  async getMultiRegionStats(regions = [], abortSignal) {
    if (!regions.length) return undefined
    const feats = await Promise.all(
      regions.map(r => this.getRegionStats(r, abortSignal)),
    )

    const scoreMax = feats
      .map(s => s.scoreMax)
      .reduce((acc, curr) => Math.max(acc, curr))
    const scoreMin = feats
      .map(s => s.scoreMin)
      .reduce((acc, curr) => Math.min(acc, curr))
    const scoreSum = feats.map(s => s.scoreSum).reduce((a, b) => a + b, 0)
    const scoreSumSquares = feats
      .map(s => s.scoreSumSquares)
      .reduce((a, b) => a + b, 0)
    const featureCount = feats
      .map(s => s.featureCount)
      .reduce((a, b) => a + b, 0)
    const basesCovered = feats
      .map(s => s.basesCovered)
      .reduce((a, b) => a + b, 0)

    return rectifyStats({
      scoreMin,
      scoreMax,
      featureCount,
      basesCovered,
      scoreSumSquares,
      scoreSum,
    })
  }

  /**
   * Fetch features for a certain region
   * @param {Region} param
   * @returns {Observable[Feature]} Observable of Feature objects in the region
   */

  getFeatures({ /* assembly, */ refName, start, end }:Region, {signal, bpPerPx}) {
    return ObservableCreate(async observer => {
      const observable2 = await this.bigwig.getFeatureStream(
        refName,
        start,
        end,
        {
          signal,
          basesPerSpan: (bpPerPx * 20) / Math.log(end - start),
        },
      )
      return observable2
        .pipe(
          mergeMap(x => from(x)),
          map(record => {
            return new SimpleFeature({
              id: record.start + 1,
              data: record,
            })
          }),
        )
        .subscribe(observer)
    })
  }
}
