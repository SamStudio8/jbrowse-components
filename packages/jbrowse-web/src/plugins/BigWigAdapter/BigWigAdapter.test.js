import { toArray } from 'rxjs/operators'

import BigWigAdapter from './BigWigAdapter'

test('adapter can fetch features from volvox.bw', async () => {
  const adapter = new BigWigAdapter({
    bigWigLocation: { localPath: require.resolve('./test_data/volvox.bw') },
  })

  const features = await adapter.getFeatures({
    refName: 'ctgA',
    start: 0,
    end: 20000,
  })
  expect(await adapter.refIdToName(0)).toBe('ctgA')
  expect(await adapter.refIdToName(1)).toBe(undefined)
  expect(await adapter.hasDataForRefName('ctgA')).toBe(true)
  expect(await adapter.hasDataForRefName('ctgB')).toBe(false)

  const featuresArray = await features.pipe(toArray()).toPromise()
  const featuresJsonArray = featuresArray.map(f => f.toJSON())
  expect(featuresJsonArray.slice(1000, 1010)).toMatchSnapshot()
})
