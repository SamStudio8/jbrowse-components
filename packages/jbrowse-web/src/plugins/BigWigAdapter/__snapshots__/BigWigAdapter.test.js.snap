// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adapter can fetch features from volvox.bw 1`] = `
Array [
  Object {
    "end": 1003,
    "score": 18,
    "start": 1002,
    "uniqueId": "1003",
  },
  Object {
    "end": 1004,
    "score": 18,
    "start": 1003,
    "uniqueId": "1004",
  },
  Object {
    "end": 1005,
    "score": 18,
    "start": 1004,
    "uniqueId": "1005",
  },
  Object {
    "end": 1006,
    "score": 19,
    "start": 1005,
    "uniqueId": "1006",
  },
  Object {
    "end": 1007,
    "score": 20,
    "start": 1006,
    "uniqueId": "1007",
  },
  Object {
    "end": 1008,
    "score": 20,
    "start": 1007,
    "uniqueId": "1008",
  },
  Object {
    "end": 1009,
    "score": 20,
    "start": 1008,
    "uniqueId": "1009",
  },
  Object {
    "end": 1010,
    "score": 20,
    "start": 1009,
    "uniqueId": "1010",
  },
  Object {
    "end": 1011,
    "score": 20,
    "start": 1010,
    "uniqueId": "1011",
  },
  Object {
    "end": 1012,
    "score": 19,
    "start": 1011,
    "uniqueId": "1012",
  },
]
`;
